doc/neps/nep-0023-backwards-compatibility.rst

```diff
@@ -19,46 +19,195 @@ processes for individual cases where breaking backwards compatibility
+
+Getting hard data on the impact of a deprecation of often difficult. Strategies
+that can be used to assess such impact include:
+
+- Use a code search engine ([1]_) or static ([2]_) or dynamic ([3]_) code
+  analysis tools to determine where and how the functionality is used.
+- Testing prominent downstream libraries against a development build of NumPy
+  containing the proposed change to get real-world data on its impact.
+- Making a change in master and reverting it, if needed, before a release. We
+  do encourage other packages to test against NumPy's master branch, so this
+  often turns up issues quickly.
+
+If the impact is unclear or significant, it is often good to consider
+alternatives to deprecations. For example discouraging use in documentation
+only, or moving the documentation for the functionality to a less prominent
+place or even removing it completely. Commenting on open issues related to it
+that they are low-prio or labeling them as "wontfix" will also be a signal to
+users, and reduce the maintenance effort needing to be spent.
+
+
+Implementing deprecations and removals
+--------------------------------------
+
+Deprecation warnings are necessary in all cases where functionality
+will eventually be removed.  If there is no intent to remove functionality,
+then it should not be deprecated either. A "please don't use this" in the
+documentation or other type of warning should be used instead.
+
+Deprecations:
+
+- shall include the version number of the release in which the functionality
+  was deprecated.
+- shall include information on alternatives to the deprecated functionality, or a
+  reason for the deprecation if no clear alternative is available.
+- shall use ``VisibleDeprecationWarning`` rather than ``DeprecationWarning``
+  for cases of relevance to end users. For cases only relevant to
+  downstream libraries, a regular ``DeprecationWarning`` is fine.
+  *Rationale: regular deprecation warnings are invisible by default; library
+  authors should be aware how deprecations work and test for them, but we can't
+  expect this from all users.*
+- shall be listed in the release notes of the release where the deprecation is
+  first present.
+- shall set a ``stacklevel``, so the warning appears to come from the correct
+  place.
+- shall be mentioned in the documentation for the functionality. A
+  ``.. deprecated::`` directive can be used for this.
+
+Examples of good deprecation warnings:
+
+.. code-block:: python
+
+    warnings.warn('np.asscalar(a) is deprecated since NumPy 1.16.0, use '
+                  'a.item() instead', DeprecationWarning, stacklevel=3)
+
+    warnings.warn("Importing from numpy.testing.utils is deprecated "
+                  "since 1.15.0, import from numpy.testing instead.",
+                  DeprecationWarning, stacklevel=2)
+
+    # A change in NumPy 1.14.0 for Python 3 loadtxt/genfromtext, slightly
+    # tweaked in this NEP (original didn't have version number).
```

Comment:
This doesn't quite reflect current practice. 

_Python_
 Note comment and stacklevel.
```
    # NumPy 1.15.0, 2017-12-10
    warnings.warn(
        "np.loads is deprecated, use pickle.loads instead",
        DeprecationWarning, stacklevel=2)
    return pickle.loads(*args, **kwargs)
```
_C_
Note comment.
```
        /* DEPRECATED 2020-05-13, NumPy 1.20 */
        if (PyErr_WarnFormat(PyExc_DeprecationWarning, 1,
                matrix_deprecation_msg, ufunc->name, "first") < 0) {
            return NULL;
        }
```

This is more howto than policy, maybe it should go elsewhere.

Reply:
I don't think I agree with these changed examples. What is the user supposed to do, read the source code or grep release notes to find the version it was introduced?

Reply:
Grep the source code is what I do, it helps if the information has a standard form recognized by a regex. I don't disagree that a NumPy version should be included in the message as well.

Reply:
Ah, you're just commenting on the comment. Okay, I'll take that over, the comment here is not part of the example, it's indicating that the example warning was changed from how it looks in the code base.

Reply:
> This is more howto than policy, maybe it should go elsewhere.

Probably useful to keep, it doesn't take up much space and examples are usually helpful.